chore: Switch to `ruint`, cleanup conversion functions and add tests

winston-h-zhang opened the pull request:
This PR swtiches the bignum crate from `crypto-bigint` to `ruint`. The motivation for this particular switch is because https://github.com/philsippl/circom-witness-rs/tree/master uses `ruint`. After considering whether to switch to `ruint` or stay with `crypto-bigint` and port the code from `circom-witness-rs`, I decided `ruint` is a bit easier to work with. 

After making the switch, I cleaned up the `from_vec_u32`, etc, conversion functions into a utils module with tests for safety.

huitseeker commented on src/witness/memory.rs:
```diff
@@ -186,7 +186,7 @@ impl SafeMemory {
     /// * `fr` - The [`U256`] field element to write.
     fn write_short(&mut self, store: &impl AsStoreRef, ptr: usize, fr: U256) -> Result<()> {
-        let num = fr.to_words()[0] as u32;
+        let num = fr.as_limbs()[0] as u32; // wtf is happening
```
what's the issue here?

winston-h-zhang replied:
I thought there might be a bug here, but after the tests passed and the constraints were satisfied, the behavior is correct

huitseeker commented on src/util.rs:
```diff
@@ -0,0 +1,92 @@
+pub fn ff_as_limbs<F: PrimeField>(f: &F) -> &[u32; 8] {
+    let binding = f.to_repr();
+    let repr: &[u8; 32] = binding.as_ref().try_into().unwrap();
+    // this doesn't work if the platform we're on is not little endian :scream:
```
It's even a bit worse than that: it assumes that the field implementation is using LE for `to_repr()`, see the tests introduced in https://github.com/lurk-lab/arecibo/pull/353/commits/fd3ae7a5e989ab62e6c128f79f5ebfae972f7fac

winston-h-zhang replied:
Yes, should I be more explicit in mentioning this issue? There's also no way to directly test within `circom-scotia`, since we can't guarantee downstream will use any particular field impl... soooo

tchataigner replied:
I guess we have two ways of addressing this:
1. Keep this as a limitation for now, document it more clearly and open an issue
2. Change the trait we deal with to `PrimeFieldBits` instead of `PrimeField` to allow for `to_le_bits` method call, ensuring correctness.

@Hanting Zhang, feel free to chose any.

huitseeker reviewed (approved):
This LGTM, just left a question

tchataigner commented on src/util.rs:
```diff
@@ -0,0 +1,93 @@
+/// Assumes little endian
+pub fn ff_as_limbs<F: PrimeField>(f: F) -> [u32; 8] {
+    let binding = f.to_repr();
+    let repr: [u8; 32] = binding.as_ref().try_into().unwrap();
```
Better error definition and propagation could be defined here. Or at least an expect giving us some context :)

tchataigner commented on src/util.rs:
```diff
@@ -0,0 +1,93 @@
+        *digit = limbs[i];
+    }
+
+    F::from_repr(repr).unwrap()
```
As above, would be great to have Rust error handling or at least an `expect`

tchataigner reviewed (dismissed):
Thanks! That looks awesome, I have few nitpicks to look at but I think we should take one of the two ways I suggested for the endianness problem.

winston-h-zhang commented:
@tchataigner Thanks for the review! I rewrote the functions to use `PrimeFieldBits` so that we no longer rely on any implementation details. This also removed the calls to `.unwrap()`, so everything worked out nicely

tchataigner reviewed (approved):
LGTM!
